From 85dfa03ec29d5e71db0f6a02d7b77df0466f7d4e Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Fri, 10 May 2024 01:11:45 +0000
Subject: [PATCH 34/40] ui/gtk: Regrabbing only on the VC which owned it before

Grabbing should be sustained for the VC which owned
the grabbing status before.

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 ui/gtk.c | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/ui/gtk.c b/ui/gtk.c
index b27d0eb4a096..c0d9b5755e75 100644
--- a/ui/gtk.c
+++ b/ui/gtk.c
@@ -897,12 +897,17 @@ static void gd_ui_show(VirtualConsole *vc)
     info.height = gdk_window_get_height(window);
     dpy_set_ui_info(vc->gfx.dcl.con, &info, false);
 
-    if (gd_is_grab_active(s)) {
-        gd_grab_keyboard(vc, "user-request-main-window");
-        gd_grab_pointer(vc, "user-request-main-window");
-    } else {
-        gd_ungrab_keyboard(s);
-        gd_ungrab_pointer(s);
+    if (s->ptr_owner && s->kbd_owner &&
+        s->ptr_owner == vc && s->kbd_owner == vc) {
+        if (gd_is_grab_active(s)) {
+            s->ptr_owner = NULL;
+            s->kbd_owner = NULL;
+            gd_grab_keyboard(vc, "user-request-main-window");
+            gd_grab_pointer(vc, "user-request-main-window");
+        } else {
+            gd_ungrab_keyboard(s);
+            gd_ungrab_pointer(s);
+        }
     }
 }
 
-- 
2.34.1

